Improve sincospi availability and implementation

sethrj opened the pull request:
I've had reports from one user building on a macOS machine with Homebrew-installed Clang, and another using a Conda-provided clang, of failures to build `Algorithms.hh`:
```
celeritas/src/corecel/math/Algorithms.hh:515:1: error: no member named '__sincospi' in the global namespace
CELER_WRAP_MATH_FLOAT_DBL_PTR_2(__, sincospi)
^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
celeritas/src/corecel/math/Algorithms.hh:399:18: note: expanded from macro 'CELER_WRAP_MATH_FLOAT_DBL_PTR_2'
        return ::PREFIX##FUNC(value, a, b);                                  \
               ~~^
<scratch space>:100:1: note: expanded from here
__sincospi
^
```

The `__sincospi` is an apple extension provided in their SDK, which isn't included by other compilers.

This PR includes two enhancements:
- A cmake "try compile" functionality to allow arbitrarily prefixed system-specific `sincospi` functions, and to correctly get that without using `__APPLE__` or implementation-defined macros
- A full and correct implementation of sincospi and sincospif from https://stackoverflow.com/a/42792940/118160 for systems that don't provide the built-ins.

sethrj committed:
Fix build with conda-supplied clang

pcanal commented on src/corecel/math/Algorithms.hh:
```diff
@@ -648,8 +648,9 @@ CELER_WRAP_MATH_FLOAT_DBL_1(, sinpi)
-#elif __APPLE__ && !defined(__CLING__)
-// Apple defines __sinpi, __sinpif, __sincospi, ...
+#elif (__MAC_OS_X_VERSION_MIN_REQUIRED > 1090) \
+    || (__IPHONE_OS_VERSION_MIN_REQUIRED > 70000)
```
iphone, really?  humm ... I guess soon enough its gpu will be strong enough to run Celeritas :) :)

sethrj replied:
I copied the condition from Apple's `math.h`... might as well be "exactly right", no? 😉

pcanal replied:
Yes but (especially because it is apple) it might change without warning :(

pcanal replied:
eg.  The version on my system is:
```
#if (defined __MAC_OS_X_VERSION_MIN_REQUIRED && __MAC_OS_X_VERSION_MIN_REQUIRED < 1090) || \
    (defined __IPHONE_OS_VERSION_MIN_REQUIRED && __IPHONE_OS_VERSION_MIN_REQUIRED < 70000)
```

sethrj replied:
Well right, but the "else" conditional is the one that has sincospi. So if the `min required` is undefined then we have the condition  0 > 1090, etc., which is a nice fallback for the macros going undefined.

pcanal commented on src/corecel/math/Algorithms.hh:
```diff
@@ -648,8 +648,9 @@ CELER_WRAP_MATH_FLOAT_DBL_1(, sinpi)
 CELER_WRAP_MATH_FLOAT_DBL_PTR_2(, sincos)
-#elif __APPLE__ && !defined(__CLING__)
-// Apple defines __sinpi, __sinpif, __sincospi, ...
+#elif (__MAC_OS_X_VERSION_MIN_REQUIRED > 1090) \
```
it looks odd to use a `__` macro, isn't there a "public" version we can use here?

sethrj replied:
Alas no, I tried. They're defined in [Availability.h](https://opensource.apple.com/source/xnu/xnu-7195.81.3/EXTERNAL_HEADERS/Availability.h.auto.html) and it says they're "for use in OS header files" but at first reading I didn't notice the "OS" — so maybe we're not allowed to use them ☹️ the fallback still works though...

The only other thing I could do is add a cmake compiler test instead but that seems like overkill.

pcanal replied:
> The only other thing I could do is add a cmake compiler test instead but that seems like overkill.

it is however the 'right' thing to do ... and it even catches other compiler/os that supports it.

sethrj replied:
Thinking about it, `__APPLE__`  and `__CLING__` are both underscores too...

sethrj committed:
Move config/version cmake and files to where they're used

Follow-on to #1325 . We don't need to "configure" the `.in` files since
they're now static.

sethrj committed:
Add a note about sincospi

sethrj committed:
Add note about where roctx comes from

sethrj committed:
Refactor sincospi definition for improved accuracy and readability

sethrj committed:
Provide true implementation of sincospi

sethrj committed:
Fix configure path

sethrj committed:
Merge remote-tracking branch 'upstream/develop' into fix-weird-compilers

sethrj committed:
Hopefully fix GCC build error

sethrj committed:
fixup! Move config/version cmake and files to where they're used

sethrj committed:
Merge remote-tracking branch 'upstream/develop' into fix-weird-compilers

pcanal commented on CMakeLists.txt:
```diff
@@ -490,7 +490,7 @@ elseif(CELERITAS_USE_HIP)
   if(ROCTX_LIBRARY)
     set(CELERITAS_HAVE_ROCTX ON)
-    target_link_libraries(celeritas_device_toolkit INTERFACE ${ROCTX_LIBRARY})
+    target_link_libraries(celeritas_device_toolkit INTERFACE "${ROCTX_LIBRARY}")
```
Why are the quote needed here?

sethrj replied:
In the unlikely case that there's a space in the path: it's a good practice.

pcanal replied:
Fair enough.  There are other place where we forgot to do this :)

pcanal commented on cmake/try-sincospi.cc:
```diff
@@ -0,0 +1,33 @@
+#    error "Expected CELERITAS_SINCOSPI_PREFIX to be defined"
+#endif
+
+#define CELER_CONCAT(PREFIX, FUNC) PREFIX##FUNC
```
If it is important that those macro stays in sync with future changes in the original, we may want to note where they come from.

pcanal commented on CMakeLists.txt:
```diff
@@ -510,6 +510,18 @@ install(TARGETS celeritas_device_toolkit
 )
 
+#----------------------------------------------------------------------------#
+if(APPLE)
```
Why limit this test to apple?  We think only Apple provides it now, but other platforms may be providing now or later without us knowing.

pcanal commented on CMakeLists.txt:
```diff
@@ -510,6 +510,18 @@ install(TARGETS celeritas_device_toolkit
 )
 
+#----------------------------------------------------------------------------#
+if(APPLE)
+  set(_prefix "__")
+  try_compile(CELERITAS_TRY_${_prefix}SINCOSPI #${CMAKE_BINARY_DIR}
+    SOURCES "${PROJECT_SOURCE_DIR}/cmake/try-sincospi.cc"
+    COMPILE_DEFINITIONS "-DCELERITAS_SINCOSPI_PREFIX=${_prefix}"
+  )
+  if(CELERITAS_TRY_${_prefix}SINCOSPI)
+    set(CELERITAS_SINCOSPI_PREFIX "${_prefix}")
+  endif()
+endif()
```
```suggestion
foreach(__prefix "" "__" )
  try_compile(CELERITAS_TRY_${_prefix}SINCOSPI #${CMAKE_BINARY_DIR}
    SOURCES "${PROJECT_SOURCE_DIR}/cmake/try-sincospi.cc"
    COMPILE_DEFINITIONS "-DCELERITAS_SINCOSPI_PREFIX=${_prefix}"
  )
  if(CELERITAS_TRY_${_prefix}SINCOSPI)
    set(CELERITAS_SINCOSPI_PREFIX "${_prefix}")
    break()
  endif()
endforeach()
```

sethrj replied:
OK, I'll expand (again) the scope of the PR...

pcanal commented on src/corecel/math/Turn.hh:
```diff
@@ -35,7 +35,13 @@ struct HalfPi
+/*!
+ * Quantity denoting a full turn.
+ *
+ * Turns are a useful way of representing angles without the historical
```
This is a good improvement.  Is it related to the issue of this PR?

sethrj replied:
Yeah, this is all about sincospi which lets you use rational values for turns.

pcanal commented on src/corecel/math/detail/Sincospi.hh:
```diff
@@ -0,0 +1,206 @@
+
+#include <cmath>
+#include <cstdint>
+
```
```suggestion
#include "corecel/Config.hh"

#ifdef CELERITAS_SINCOSPI_PREFIX
#error "sincospi is provided natively on this platform. Do not include Sincospi.hh"
#endif
```

sethrj replied:
What if we want to test the implemented functions? You should only `#error` to prevent further build errors...

pcanal replied:
The problem is that if a user 'inadvertently' use this, there will be duplicate symbols or worse simple silent break of ODR (eg. if all the sincospi get inlined).  If one wants to test those function on a platform that already has `sincospi` additional care/gymnastic is anyway needed and it might make sense for that test to have to explicitly `#undef` the macro.

sethrj replied:
No there won't; the `detail` functions in this header are uniquely defined. The only potential ODR violation will be if `::celeritas::sincospi` (for example) calls one implementation in one translation unit and another implementation in another. No macro will fix that, and since they're forceinlined it probably won't double-define anyway.

pcanal replied:
> since they're forceinlined it probably won't double-define anyway.

That the worst case scenario (it silent weird result because the 2 implementation give slightly different result).

> No macro will fix that

Sort of.  I indeed completely missed that this header is defining `sincospi_impl` rather than`sincospi`.  So indeed it would take a corrupted build using 2 distinct config header file to trigger the problem.

So we can leave it as is.

sethrj committed:
Fix missing HD decorator

sethrj committed:
Expand cospi testing

sethrj committed:
Shortcut try-compile on subsequent builds and fix support for empty prefix
